### PR TITLE
Add ci vis links to linux unit and e2e tests

### PR DESCRIPTION
**Pull request by pducolin**
### What does this PR do?

* add task to generate CI Visibility links from a Gitlab job
* use the task in e2e tests
* use the task in linux unit tests

### Motivation

ADXT-288

This will help investigations on failing jobs: with a quick link to main's job execution, we can quickly determine if a failure is a flake or it is due to the branch changes

### Additional Notes

This is a rework of https://github.com/DataDog/datadog-agent/pull/26098 that was reverted due to a bug in handling job names. The original PR was creating the annotations file from the job's script, using `sed` to replace job id and job name in a template json file. Now we generate the json file from Python.

External links are a Gitlab feature that allows to define extra links in a json file that contains one or more annotations. Those links are rendered at job's completion under artifacts.

image
  
More details: https://docs.gitlab.com/ee/ci/yaml/artifacts_reports.html#external_link

### Describe how to test/QA your changes

A full pipeline execution with e2e jobs was triggered at [#36146446](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/pipelines/36146446)

Note: ensure you are in org2 when opening links

### CODE REVIEW

**Comment by pducolin** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

**Comment by pducolin** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

**Comment by pducolin** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

**Comment by pducolin** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

File: tasks/__init__.py [hunk 1]
```diff
@@ -53,6 +54,8 @@
 from tasks.components import lint_components, lint_fxutil_oneshot_test
 from tasks.custom_task.custom_task import custom__call__
 from tasks.fuzz import fuzz
+
+# from tasks.gitlab_helpers import pippo
```
Comment (amenasria) [inline tasks/__init__.py:58]:
```suggestion
```
❓ question: ‏Feels like a leftover, is it on purpose?

File: tasks/__init__.py [hunk 2]
```diff
@@ -129,6 +137,7 @@
 ns.add_task(get_impacted_packages)
 ns.add_task(get_modified_packages)
 ns.add_task(send_unit_tests_stats)
+# ns.add_task(pippo)
```
Comment (amenasria) [inline tasks/__init__.py:140]:
```suggestion

```

**❓ question:** ‏Feels like a leftover, is it on purpose?

File: tasks/gitlab_helpers.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        "@ci.pipeline.name": "DataDog/datadog-agent",
+    }
+    return f"{CI_VISIBILITY_URL}?query={quote(to_query_string(query_params))}"
+
+
+def to_query_string(params: dict):
+    query = ""
+    for key, value in params.items():
+        query += f"{key}:{value} "
+    return query
```
Comment (amenasria) [inline tasks/gitlab_helpers.py:81]:
```suggestion
    return " ".join(f"{k}:{v}" for k, v in params.items())
```

**💬 suggestion:** ‏Same code just a bit shorter, feel free to ignore if you don't like it !

File: tasks/gitlab_helpers.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    Generated file
+    """
+    ci_job_id = os.environ.get("CI_JOB_ID")
+    if ci_job_id is None:
+        print("CI_JOB_ID is not set, this task can run only on Gitlab, skipping...")
+        return
+    ci_job_name = os.environ.get("CI_JOB_NAME")
+    if ci_job_name is None:
+        print("CI_JOB_NAME is not set, this task can run only on Gitlab, skipping...")
+        return
```
Comment (amenasria) [inline tasks/gitlab_helpers.py:28]:
```suggestion
    if ci_job_id is None:
        print(color_message("CI_JOB_ID is not set, this task can run only on Gitlab, skipping...", Color.RED))
        return
    ci_job_name = os.environ.get("CI_JOB_NAME")
    if ci_job_name is None:
        print(color_message("CI_JOB_NAME is not set, this task can run only on Gitlab, skipping...", Color.RED))
        return
```

**💬 suggestion:** ‏

**Review by amenasria — APPROVED** [review]
Left a few comments otherwise LGTM ! 


**💭 thought:** ‏The functions in `tasks/gitlab_helpers.py` are short enough to be tested imo, if you have time, could you please add unit tests for them?

File: tasks/__init__.py (hunk 1 above)
Comment (pducolin) [reply tasks/__init__.py]: Ah thank you for the catch

**Review by sblumenthal — APPROVED** [review]

File: tasks/gitlab_helpers.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+@task
+def generate_ci_visibility_links(_ctx, output: str | None):
+    """
+    Generate links to CI Visibility for the current job
+    usage deva gitlab.generate-ci-visibility-links
+    Generated file
+    """
+    ci_job_id = os.environ.get("CI_JOB_ID")
+    if ci_job_id is None:
+        print(color_message("CI_JOB_ID is not set, this task can run only on Gitlab, skipping...", Color.RED))
```
Comment (FlorentClarret) [inline tasks/gitlab_helpers.py:23]: you're missing the imports

File: tasks/gitlab_helpers.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+@task
+def generate_ci_visibility_links(_ctx, output: str | None):
+    """
+    Generate links to CI Visibility for the current job
+    usage deva gitlab.generate-ci-visibility-links
+    Generated file
+    """
+    ci_job_id = os.environ.get("CI_JOB_ID")
+    if ci_job_id is None:
```
Comment (FlorentClarret) [inline tasks/gitlab_helpers.py:22]:
```suggestion
    if not ci_job_id:
```

More pythonic and would also catch empty strings

File: tasks/gitlab_helpers.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        print(color_message("CI_JOB_ID is not set, this task can run only on Gitlab, skipping...", Color.RED))
+        return
+    ci_job_name = os.environ.get("CI_JOB_NAME")
+    if ci_job_name is None:
```
Comment (FlorentClarret) [inline tasks/gitlab_helpers.py:26]:
```suggestion
    if not ci_job_name:
```

**Review by FlorentClarret — APPROVED** [review]
Left some minor comments but LGTM!

**Review by L3n41c — COMMENTED** [review]
I tried to run the `containers` e2e tests on the [pipeline](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/pipelines/36300392) of this PR.
I had a look at the [`new-e2e-containers: [--run TestECSSuite]` job](https://gitlab.ddbuild.io/DataDog/datadog-agent/-/jobs/537929669) as an example.

Although I can see the external links on the page of the job, the corresponding CI visibility page shows nothing matching the queries:
* [CI Visibility: This job instance](https://app.datadoghq.com/ci/pipeline-executions?query=ci_level%3Ajob%20%40ci.job.id%3A537929669%20%40ci.pipeline.name%3ADataDog%2Fdatadog-agent&agg_m=count&agg_m_source=base&agg_t=count&fromUser=false&index=cipipeline&start=1717491519540&end=1718096319540&paused=false)
* [CI Visibility: This job on main](https://app.datadoghq.com/ci/pipeline-executions?query=ci_level%3Ajob%20%40ci.job.name%3A%22new-e2e-containers%3A%20%5B--run%20TestECSSuite%5D%22%20%40git.branch%3Amain%20%40ci.pipeline.name%3ADataDog%2Fdatadog-agent&agg_m=count&agg_m_source=base&agg_t=count&fromUser=false&index=cipipeline&start=1717491561237&end=1718096361237&paused=false)

**Review by amenasria — APPROVED** [review]
Thanks for adding some unit tests ! 🫰

**Pull request merged**